OGRwkbGeometryType using non ISO C compliant values

## Expected behavior and actual behavior.

I was using gdal library in my source code and compiling it with -Wpedantic and came across an issue when including ogr_core.h.
```
In file included from /usr/include/gdal/ogr_srs_api.h:35,
                 from gedisimulator/gediIO.c:17:
/usr/include/gdal/ogr_core.h:399:19: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  399 |     wkbPoint25D = 0x80000001, /**< 2.5D extension as per 99-402 */
      |                   ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:400:24: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  400 |     wkbLineString25D = 0x80000002, /**< 2.5D extension as per 99-402 */
      |                        ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:401:21: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  401 |     wkbPolygon25D = 0x80000003, /**< 2.5D extension as per 99-402 */
      |                     ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:402:24: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  402 |     wkbMultiPoint25D = 0x80000004, /**< 2.5D extension as per 99-402 */
      |                        ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:403:29: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  403 |     wkbMultiLineString25D = 0x80000005, /**< 2.5D extension as per 99-402 */
      |                             ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:404:26: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  404 |     wkbMultiPolygon25D = 0x80000006, /**< 2.5D extension as per 99-402 */
      |                          ^~~~~~~~~~
/usr/include/gdal/ogr_core.h:405:32: warning: ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]
  405 |     wkbGeometryCollection25D = 0x80000007 /**< 2.5D extension as per 99-402 */
```

## Steps to reproduce the problem.

```
sudo apt-get install libgdal-dev -y
echo "#include <ogr_core.h>" > test.c
gcc -c test.c -Wpedantic $(gdal-config --cflags)
```

## Operating system

Ubuntu 18.04.4 LTS Bionic 64 bit

## GDAL version and provenance

2.2.3 version from ubuntu official repository

--- a/ogr/ogr_core.h
+++ b/ogr/ogr_core.h
@@ -391,6 +391,13 @@ typedef int OGRBoolean;
 /*      ogr_geometry.h related definitions.                             */
 /* -------------------------------------------------------------------- */
 
+#if defined(HAVE_GCC_DIAGNOSTIC_PUSH) && __STDC_VERSION__ < 202311L
+/* wkbPoint25D and friends cause warnings with -Wpedantic prior to C23. */
+/* Cf https://github.com/OSGeo/gdal/issues/2322 */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#endif
+
 /**
  * List of well known binary geometry types.  These are used within the BLOBs
  * but are also returned from OGRGeometry::getGeometryType() to identify the
@@ -511,6 +518,10 @@ typedef enum
     wkbGeometryCollection25D = 0x80000007 /**< 2.5D extension as per 99-402 */
 #endif
 } OGRwkbGeometryType;
+
+#if defined(HAVE_GCC_DIAGNOSTIC_PUSH) && __STDC_VERSION__ < 202311L
+#pragma GCC diagnostic pop
+#endif
 
 /* clang-format off */
 /**

